### PR TITLE
aarch64: pm_cpu_ops: Introduce pm_cpu_ops subsystem

### PR DESCRIPTION
**Pull request by carlocaione**
AArch64 has support for PSCI. This is especially useful for SMP because
PSCI is used to power on the secordary cores.

When the PSCI driver was introduced in Zephyr it was designed to rely on
a very PSCI-centric subsystem / interface.

There are two kinds of problems with this choice:

1. PSCI is only defined for the non-secure world and it is designed to
   boot CPU cores into non-secure state (that means that PSCI is only
   supposed to work if Zephyr is running in non-secure state)

2. There can be other ways or standards used to start / stop a core
   different from PSCI

This patch is trying to fix the original wrong assumption by making the
interface / subsystem a generic one, called 'pm_cpu_ops', and using PSCI
only as an actual driver that is a user of this new interface /
subsystem.

For now the new subsystem is only exposing two methods: cpu_on and
cpu_off, others will probably follow according to the needs.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

### CODE REVIEW

File: drivers/pm_cpu_ops/Kconfig [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+	help
+	  Enable CPU power management driver configuration
+
+if PM_CPU_OPS
+
+module = PM_CPU_OPS
+module-str = pm_cpu_ops
+source "subsys/logging/Kconfig.template.log_config"
+
+source "drivers/pm_cpu_ops/Kconfig.psci"
```
Comment (sandeepbrcm) [inline drivers/pm_cpu_ops/Kconfig:17]: something is still coupling with psci.  I think we should have global ops structure to register the ops. User / API need not worry about the underlying mechanism (psci or someother). For example RISC-V will register its SBI ops and someone may register local functions.

File: tests/arch/arm64/arm_psci/src/main.c [hunk 2]
```diff
@@ -24,12 +25,8 @@ void test_psci_func(void)
 		       PSCI_VERSION_MINOR(ver) < 2),
 		       "Wrong PSCI firware version");
 
-	/* This should return 0: (one core in the affinity instance is ON) */
-	ret = psci_affinity_info(psci, 0, 0);
-	zassert_true(ret == 0, "Wrong return code from psci_affinity_info");
-
 	/* This should return -PSCI_RET_ALREADY_ON that is mapped to -EINVAL */
-	ret = psci_cpu_on(psci, 0, 0);
+	ret = pm_cpu_on(psci, 0, 0);
```
Comment (sandeepbrcm) [inline tests/arch/arm64/arm_psci/src/main.c:29]: To operate on cpu what the user has to do ?
Get the device handle to a "pm compatible device"  which can be PSCI or something else.
and operate on it.
Instead should it just operate on the apis like `pm_cpu_ops.cpu_on(cpuid)`

The change required will be to maintain `pm_cpu_ops`  struct and provide registration api. 
The ops provider may or may not be a device. choosing the ops provider is a platform config choice. CONFIG_PSCI etc. The is conventional approach.

I want to see pros in this new approach as well if we have to deviate from the conventional approach. Thoughts?

File: CODEOWNERS [hunk 3]
```diff
@@ -244,7 +244,7 @@
 /drivers/peci/                            @albertofloyd @franciscomunoz @scottwcpg
 /drivers/pinmux/*hsdk*                    @iriszzw
 /drivers/pinmux/*it8xxx2*                 @ite
-/drivers/psci/                            @carlocaione
+/drivers/pm_cpu_ops/                      @carlocaione
```
Comment (sandeepbrcm) [inline CODEOWNERS:247]: now this is a generic cpu ops subsystem . architecture agnostic.  +1

File: drivers/pm_cpu_ops/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library()
+
+zephyr_library_sources_ifdef(CONFIG_ARM_PSCI	pm_cpu_ops_psci.c)
+zephyr_library_sources_ifdef(CONFIG_USERSPACE	pm_cpu_ops_handlers.c)
```
[the remark this answers is not in the document]
Comment (carlocaione) [reply drivers/pm_cpu_ops/CMakeLists.txt:6]: Not sure what you mean. The handlers are selectively compiled on `CONFIG_USERSPACE`. Where you referring to something different?

File: drivers/pm_cpu_ops/Kconfig (hunk 1 above)
Comment (carlocaione) [reply drivers/pm_cpu_ops/Kconfig:17]: What you think is coupling with PSCI? The `pm_cpu_ops` subsystem is now totally agnostic. The PSCI implementation is only using this agnostic interface. Here we are sourcing the PSCI Kconfig file.

File: tests/arch/arm64/arm_psci/src/main.c (hunk 2 above)
Comment (carlocaione) [reply tests/arch/arm64/arm_psci/src/main.c:29]: > To operate on cpu what the user has to do ?
> Get the device handle to a "pm compatible device" which can be PSCI or something else.
> and operate on it.

Correct.

> Instead should it just operate on the apis like `pm_cpu_ops.cpu_on(cpuid)`
> 
> The change required will be to maintain `pm_cpu_ops` struct and provide registration api.
> The ops provider may or may not be a device. choosing the ops provider is a platform config choice. CONFIG_PSCI etc. The is conventional approach.
> 
> I want to see pros in this new approach as well if we have to deviate from the conventional approach. Thoughts?

Another hybrid approach can be something like you proposed but leveraging the DTS `enable-method` of the cpu nodes property to register the proper driver / mechanism.

File: drivers/pm_cpu_ops/CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (sandeepbrcm) [reply drivers/pm_cpu_ops/CMakeLists.txt:6]: `CONFIG_USERSPACE `  is not relevant to cpu_ops right?

[the remark this answers is not in the document]
Comment (carlocaione) [reply drivers/pm_cpu_ops/CMakeLists.txt:6]: well, it depends if you want to give userspace the possibility to shut down a core for example. I'm not strongly opinionated on this so I just decided to create the handlers.

[the remark this answers is not in the document]
Comment (carlocaione) [reply drivers/pm_cpu_ops/CMakeLists.txt:6]: I'll take this out in the next respin.

**Comment by carlocaione** [comment]
@sandeepbrcm I think I have addressed your suggestions. The interface now is not tied to a device but it's a system-wide interface.

File: include/drivers/pm_cpu_ops.h [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * @defgroup power_management_cpu_api CPU Power Management
+ * @{
+ */
+
+/**
+ * @brief Power down the calling core
+ *
+ * This call is intended foruse in hotplug. A core that is powered down by
```
Comment (jharris-intel) [inline include/drivers/pm_cpu_ops.h:31]:
```suggestion
 * This call is intended for use in hotplug. A core that is powered down by
```

File: include/drivers/pm_cpu_ops.h [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+ * into the calling supervisory software or have been previously powered down
+ * with a cpu_off call
+ *
+ * @param cpuid CPU id to power on
+ * @param entry_point Address at which the core must commence execution
+ *
+ * @retval 0 on success, a negative errno otherwise
+ * @retval -ENOTSUP If the operation is not supported
+ */
+int pm_cpu_on(unsigned long cpuid, unsigned long entry_point);
```
Comment (jharris-intel) [inline include/drivers/pm_cpu_ops.h:52]:
```
typedef void (entry_f)(void);
[...]
int pm_cpu_on(unsigned long cpuid, entry_f *entry_point);
```

...or at the very least `uintptr_t` please for an address.

File: include/drivers/pm_cpu_ops.h (hunk 5 above)
Comment (carlocaione) [reply include/drivers/pm_cpu_ops.h]: ACK

File: include/drivers/pm_cpu_ops.h (hunk 6 above)
Comment (carlocaione) [reply include/drivers/pm_cpu_ops.h]: Agree. Fixed (using `uintptr_t`).

**Review by npitre — APPROVED** [review]

**Comment by MrVan** [comment]
> AArch64 has support for PSCI. This is especially useful for SMP because
> PSCI is used to power on the secordary cores.
> 
> When the PSCI driver was introduced in Zephyr it was designed to rely on
> a very PSCI-centric subsystem / interface.
> 
> There are two kinds of problems with this choice:
> 
> 1. PSCI is only defined for the non-secure world and it is designed to
>    boot CPU cores into non-secure state (that means that PSCI is only
>    supposed to work if Zephyr is running in non-secure state)
> 2. There can be other ways or standards used to start / stop a core
>    different from PSCI
> 
> This patch is trying to fix the original wrong assumption by making the
> interface / subsystem a generic one, called 'pm_cpu_ops', and using PSCI
> only as an actual driver that is a user of this new interface /
> subsystem.
> 
> For now the new subsystem is only exposing two methods: cpu_on and
> cpu_off, others will probably follow according to the needs.
> 

So you are going to support spin table for ARM64? Should we force PSCI for ARM64?

**Comment by carlocaione** [comment]
> So you are going to support spin table for ARM64?

I'm not personally interested in spin tables but in that case we can hook that up to this interface.

> Should we force PSCI for ARM64?

You do not always want PSCI on ARM64. For example you do not want that when you do not have a capable firmware running in EL2 or EL3 or when booting in secure mode. Or in general you can have other ways to bring your CPUs up (IIRC @jharris-intel is using some custom method instead of PSCI). So, no, the idea is not to force that.

File: CODEOWNERS (hunk 3 above)
Comment (MrVan) [inline CODEOWNERS:247]: pm_cpu_ops seems not a good name to me. PSCI will be used for power management also, such as suspend/resume.
I would keep psci still as it is or like Linux Kernel use a firmware directory.

Comment (carlocaione) [reply CODEOWNERS:247]: This is exactly the opposite of what we are trying to do here. PSCI is just one way to perform CPU power management, but it's not the only way. To be able to support multiple implementations we have to create a "generic" interface for the CPU power management. PSCI is then using this generic interface to provide its services.

For example in `drivers/gpio` you have all the drivers using / implementing the GPIO interface, you do not have a directory for each implementation. The same here for `pm_cpu_ops`, in this directory you will find all the different users of this interface, including PSCI but possibly others as well.

About the name: `pm_cpu_ops` is literally taken from power management CPU operations. That can be a bit ugly but definitely better than `firmware` that is way too generic and broad.

About suspend/resume you are right but this is not something we currently support so I decided for now to leave those out. We can add the hooks to the interface when we will have something using them.

**Review by sandeepbrcm — APPROVED** [review]

File: CODEOWNERS (hunk 3 above)
Comment (sandeepbrcm) [reply CODEOWNERS:247]: Re, the name `pm_cpu_ops`  you may use if anything more abstract pops up. So far I think this is appropriate. And the list of ops can be extended to cover all `known` cpu ops later from all architecture.

**Pull request merged**